atomix/cluster/src/test/java/io/atomix/raft/ControllableRaftContexts.java

```diff
@@ -359,4 +360,66 @@ private void updateAndVerifyLeaderTerm(final RaftContext s) {
       }
     }
   }
+
+  // If a node is in CandidateRole, then it will update the term. But until the election is
+  // completed, there is no leader at that term.
+  public boolean hasLeaderAtTheLatestTerm() {
+    final var currentTerm =
+        raftServers.values().stream().map(RaftContext::getTerm).max(Long::compareTo).orElseThrow();
+    return leadersAtTerms.get(currentTerm) != null;
+  }
+
+  boolean hasCommittedAllEntries() {
+    return raftServers.values().stream()
+        .noneMatch(
+            s -> {
+              final var lastCommittedEntry = getLastCommittedEntry(s);
+              final var lastUncommittedEntry = getLastUncommittedEntry(s);
+
+              return lastUncommittedEntry != null
+                  && !lastUncommittedEntry.equals(lastCommittedEntry);
+            });
+  }
+
+  boolean hasReplicatedAllEntries() {
+    return raftServers.values().stream().map(this::getLastUncommittedEntry).distinct().count() == 1;
+  }
+
```

Comment:
:question: Not sure why we check here whether the count is equal to one.

Reply:
So my current assumption is because:

it could return either null and than it is equal for all OR could also be that all have one uncommitted :thinking: 

If one has null and one uncommitted it is 2 and not replicated to everyone.

Reply:
Yes. That is correct.